### ISSUE TITLE
Loud noise if Input is disabled in ASIO4ALL while still selected in Jamulus

### ISSUE THREAD

**Issue by ann0see** [jamulussoftware/jamulus#796]
If you select an input in ASIO4ALL, set it in Jamulus and later disable it in ASIO4All (and obviously can't in Jamulus since the in/output dialog is hidden) this results in loud noise on the unavailable input. This might be related to some buffers not being cleared.

Steps to reproduce:

1. Select an (internal) soundcard with multiple inputs: e.g. enable your integrated mic(s) and your headphone mic.
2. Set the left channel in Jamulus to your internal left mic and set your right channel to the mic of your headphones
3. Disable one of the microphone inputs in ASIO4ALL --> the input will no longer be available in Jamulus and you will hear a terrible loud noise on the input you disabled.
4. Enable the input: the sound disappears. 

Expected behaviour:

Debatable.

1. Give a warning that the inputs changed and tell the user to select the correct inputs/outputs in Jamulus;
2.  Optional: if now there's only one available, automatically select it (or ask the user if he wants to set this)

**Comment by corrados** [comment]
I cannot reproduce the loud noise but when I enable/disable inputs/outputs in ASIO4All, the Jamulus software freezes and then crashes. So I think it makes sense to first try to fix the issue with the crash.

**Comment by ann0see** [comment]
> the Jamulus software freezes and then crashes

oh. That’s even worse. Luckily it didn’t for me.

**Comment by nico-dr** [comment]
My expierence is that this is related to exclusive access (the windows default) to the audio hardware 

I think you can workaround these kind of freeze by disabling the exclusive access in the windows settings.

![image](https://user-images.githubusercontent.com/76488290/103209935-8ed5b880-4904-11eb-84c6-d64bc527fd8a.png)

**Comment by ann0see** [comment]
Do you know if there are any drawbacks once you disable exclusive access?

**Comment by corrados** [comment]
> I cannot reproduce the loud noise but when I enable/disable inputs/outputs in ASIO4All, the Jamulus software freezes and then crashes.

I just debugged the issue. The problem is that the function ASIOCreateBuffers() in CSound::Init() suddenly blocks and does not return. When this happens, I then run into the MutexDriverReinit.lock() of CClient::OnSndCrdReinitRequest with the GUI thread and the GUI gets stuck.

The problem is that the function ASIOCreateBuffers() is not under my control. So if that function suddenly blocks, I have a problem which I cannot solve inside the Jamulus source code. The only chance is to somehow predict that this function will most likely block. But the question is, how to figure that out? I am running out of ideas...

**Comment by ann0see** [comment]
At least for me, the input/output selection in Jamulus disappears if I select the wrong device. Maybe you can somehow detect this?

**Comment by ann0see** [comment]
Ok. Now I also got a freeze. Not 100% sure what I did wrong, but while changing the in/outputs in ASIO4ALL Jamulus indeed froze.

**Comment by corrados** [comment]
I think I now know what goes wrong. I introduced this issue in the last release version. I'll fix this for the next release.

**Comment by ann0see** [comment]
@corrados if I can help somehow with this I'd be glad to do so. 
Warning: I don't know C++ well.

**Comment by ann0see** [comment]
@pljones 

Do you have any idea on this bug? Since Volker left, I assume you're the only one who could help here. 

I've now put the ASIO4ALL issues on my agenda and am trying to understand where all the problems come from. This bug is probably not making the issue better.

**Comment by pljones** [comment]
Sorry, I've not really ever looked at the audio code, other than to link the recorder to the mixer.

**Comment by ann0see** [comment]
Hmm. That's a pitty. Not sure if we can/should e-mail corrados or find somebody who might have an idea what's wrong here.

**Comment by pljones** [comment]
Diagnostic questions:
- Does this happen regardless of the underlying device in ASIO4ALL?
- Does this happen with any other driver than ASIO4ALL?

If yes to either of the above, it's not a Jamulus issue and quite possibly not something Jamulus can fix: it may not "know" enough about what's happening at the ASIO driver interface level.  If yes, then some work in that area to pin down what goes wrong is likely needed.

**Comment by ann0see** [comment]
> Does this happen with any other driver than ASIO4ALL?

No idea about this. At least no other ASIO driver I know allows changing input/output devices on runtime.

> Does this happen regardless of the underlying device in ASIO4ALL?

I think so. At least I can reproduce this error with my sound card and corrados can too (and he probably also knows how to solve it. I will try an old version of Jamulus and check if this bug can reproduced.)

**Comment by ann0see** [comment]
By the way, there is an API for ASIO4ALL the developer just sent to me: https://www.asio4all.org/?page_id=349
Although I thought about this, I don't think we can use much of this in Jamulus. Some things concerning ASIO4ALL can maybe be solved (this issue if it is Jamulus related), others should be improved in A4A.

**Comment by ann0see** [comment]
I can not reproduce this error in v3.5.12. If I disable a device the selection in Jamulus doesn't disappear.

**Comment by ann0see** [comment]
@hselasky I know you develop HPS Jam and probably also know something about the ASIO API. I will look through the commits and check which one introduced this bug, but maybe you have an idea here?

I think, this bug (especially the freeze; but I think that they're related) must be fixed before a new release is made.

**Comment by hselasky** [comment]
Haven't seen this particular issue.

**Comment by ann0see** [comment]
Thanks for your answer! So I assume it's really a Jamulus only problem. I'm going to look at d8a5c6f and have a few thoughts especially about https://github.com/jamulussoftware/jamulus/commit/d8a5c6fdc7ec9bceaf6fb8569dce1a2962c64597#diff-91bd7e280760d71193396d52e45966013010d348eebeb2ab26bd5a0f66b58393R73

Although I haven't tested if this commit introduced this bug, it would be quite logical that it did:

If ASIO4ALL inputs change, the driver doesn't get reloaded anymore after this commit. Therefore Jamulus might try to access a output/input which isn't there anymore --> crash or the behavior I described in my first post. Do you think this is possible (as I said, I'm not a C++/ASIO expert...)

**Comment by hselasky** [comment]
You could give HpsJAM a quick spin with ASIO4ALL. It's at my github page and see if the issue is the same. It might also be one of those unsupported sample rate formats! I know I made a big patch which never got accepted, so the answer may be in there.

**Comment by ann0see** [comment]
Ok. Thanks! In the meantime, I did a few tests and commented the if in line 73 windows/sound.cpp and the issue is gone! So my thesis seemed to be right.

I will have a few more thoughts on this, since this might break some other functionality.

@hselasky Concerning the unsupported sample rate issues, maybe that's worth another issue with smaller changes. I don't exactly remember why your PR was rejected, but I think it was due to more processing power being needed? If there are other improvements, probably it's worth another try ;-).

**Comment by softins** [comment]
> Ok. Thanks! In the meantime, I did a few tests and commented the if in line 73 windows/sound.cpp and the issue is gone! So my thesis seemed to be right.

Looking at the commit history for `windows/sound.cpp` between 3.6.1 and 3.6.2:
```
git log --patch r3_6_1..r3_6_2 windows/sound.cpp
```
I think the most likely-looking culprit is d8a5c6f. This seems to agree with your comment above, although the line number doesn't quite match.

**Comment by ann0see** [comment]
@softins Probably I just misread the line. Any idea except removing the if statement to solve this bug? We must somehow detect if the ASIO driver options (especially the inputs/outputs) changed. The (hypothetical) problem here: we can't rely on the number of inputs since ASIO4ALL might show the same number of inputs/outputs although they're changed. Also, they might be named the same. We could only rely on an UID, but I doubt this is possible.

**Comment by pljones** [comment]
Yes, that change (https://github.com/jamulussoftware/jamulus/commit/d8a5c6fdc7ec9bceaf6fb8569dce1a2962c64597) looks dubious -- DAWs I've used will close the sound device completely whilst the audio device control panel is open.  I'd expect Jamulus to stop reading and writing the sound card whilst the control panel is open and then re-attach, if it can, when it closes.

What this means is something _may_ need to change in how the connection to the server is handled -- i.e. I'm not sure if we'd need to feed silence and route received audio to nowhere just to keep things running.

**Comment by ann0see** [comment]
>  I'm not sure if we'd need to feed silence and route received audio to nowhere just to keep things running.

I'm not sure if this is needed: What's the difference between a short cut in audio because of a driver change between a short cut in network traffic which just occurs randomly? Jamulus stays "connected".

And yes, I think, there are some problems related to the unclean reload of the ASIO interface (other bugs I've came accross are maybe also related to this: Sometimes red crosses are (at least from my impression) might be related to Jamulus blocking itself)

**Comment by softins** [comment]
> @softins Probably I just misread the line. Any idea except removing the if statement to solve this bug? We must somehow detect if the ASIO driver options (especially the inputs/outputs) changed.

I wasn't sure exactly which `if` you meant by line 73, as there were a couple nearby.

I'm afraid I don't have any knowledge of ASIO, nor a dev environment for PC at present, so I can't comment in detail.

**Comment by pljones** [comment]
This:
```
if ( strCurDevName.compare ( strDriverNames[iDriverIdx] ) != 0 )
```

Basically, it's saying "unless the sound driver changed, don't reset the channel mapping".  Of course, with ASIO4ALL, that doesn't make sense, as the channel mapping in ASIO4ALL could have changed, so Jamulus needs a reset.  Indeed, with many multi-channel or multi-card ASIO drivers, it doesn't make sense.

It's _dangerous_ - you can drive bad data through the system - to leave audio processing running whilst changing hardware mapping.  You could cause physical damage to equipment and, in worst case, to hearing.

**Comment by ann0see** [comment]
So what would be your Suggestion? The problem here is that we need to have audio right before and right after the change of the inputs outputs because people do not know which device to choose in ASIO4ALL since there are ambiguities. I for example have two microphones named exactly the same in ASIO4ALL. Therefore you must try and listen what works.

**Comment by ann0see** [comment]
@pljones would you remove the line totally (at least for a short term fix)?
Should we check if the driver name is ASIO4ALL and then reload the channel mapping or would you suggest to do a big change now or in near future?

**Comment by pljones** [comment]
Short term, the `if { ... }` block that was inserted with that patch could be reverted -- lines windows/sound.cpp 72-81 reverting to just the inner two lines (plus comments) that were there before.

**Comment by ann0see** [comment]
Ok. I can open a PR on this, if you want?

**Comment by ann0see** [comment]
I think in general the code around this line is a bit dubious:

 strStat will never be empty even if the settings in ASIO4ALL are set correctly. I think we should re-start Jamulus if we come here. This might also solve https://github.com/jamulussoftware/jamulus/issues/872 which I assume is also related to the code here.

**Comment by pljones** [comment]
First, have you tried the fix?  Does it help?

Regarding `strStat`:
https://github.com/jamulussoftware/jamulus/commit/d8a5c6fdc7ec9bceaf6fb8569dce1a2962c64597#diff-91bd7e280760d71193396d52e45966013010d348eebeb2ab26bd5a0f66b58393R112
> If the return string is empty, the device can be used, otherwise the error message is returned.

So it _should_ always be empty -- if the selected device (that's the ASIO _driver_) is capable.

We shouldn't need to re-start Jamulus if the device changes but I think we need to effectively disconnect audio inputs and outputs whilst changing the device is possible.  And that has an effect on what we're sending to the server, currently.

**Comment by ann0see** [comment]
> So it should always be empty -- if the selected device (that's the ASIO driver) is capable.

It isn't: check line 76 and line 92. (not sure if these are the correct lines now). 
Let's think of a user wrongly setting a 44.1kHz device in ASIO4ALL. This leads to an error message and asks if you'd like to open the driver setup window.
The driver window is opened, the user configures the in/outputs correctly and clicks on yes on the message asking if he's done with the driver setup. According to my understanding of the code no checks are made once the setup is finished --> strStat will still not be empty even if the setup was correct --> leads to the "Jamulus could not be started because of Audio interface issues" error message (at least that was a finding from a few weeks ago).

**Comment by pljones** [comment]
`strStat` gets set by calling `CheckDeviceCapabilities()` on line 67.  It doesn't exist until that call has been made and doesn't change after that call has been made.  So it's reset every time `LoadAndInitializeDriver()` is called - you can't get to the `if` that checks it without creating it afresh.

**Comment by ann0see** [comment]
Discussion should continue in #727, #872 or a new issue.

**Comment by ann0see** [comment]
Found another (maybe related) issue: If you select a device which doesn't support 48kHz and a device which does in A4A Jamulus hangs. I think we need to investigate this issue further. Not sure if that's a blocker for the next release.

**Comment by ann0see** [comment]
I think I'll close this again and raise a new issue.

**Comment by ann0see** [comment]
> This:
> 
> ```
> 
> if ( strCurDevName.compare ( strDriverNames[iDriverIdx] ) != 0 )
> 
> ```
> 
> 
> 
> Basically, it's saying "unless the sound driver name changed, don't reset the channel mapping".  Of course, with ASIO4ALL, that doesn't make sense, as the channel mapping in ASIO4ALL could have changed, so Jamulus needs a reset.  Indeed, with many multi-channel or multi-card ASIO drivers, it doesn't make sense.
> 
> 
> 
> It's _dangerous_ - you can drive bad data through the system - to leave audio processing running whilst changing hardware mapping.  You could cause physical damage to equipment and, in worst case, to hearing.

@pljones

Anything which we need to change for r 3.7.0 here?

**Comment by pljones** [comment]
For 3.7.0, I think the `if` above should be reverted.  I think it's a safe change to make.

For 3.7.1, I think:
* if the selected driver is changed (i.e. if the index into the list changes)
* when the driver control panel is requested and until it's closed

we should send silence and drop incoming audio (assuming there's a server connection, of course).

Hopefully, doing the above protects anyone connected to the server (including _this_ client) from corrupted audio due to the change.
